### PR TITLE
remove --verbose from rock entrypoint

### PR DESCRIPTION
**Pull request by linostar**
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----

This was triggered by the existing pebble bug at  https://github.com/canonical/pebble/issues/339.

Action to be taken in Rockcraft:
- [x] remove `--verbose` from the rock’s entrypoint. That should always have been the default behavior, also because pebble doesn’t have a `--quiet` option. Those who want verbose mode, can always still pass --verbose at container deployment time.

### CODE REVIEW

**Comment by linostar** [comment]
@cjdcordeiro This should be ready for review.

**Comment by tigarmo** [comment]
@linostar @cjdcordeiro there are many spread failures due to this change that need to be addressed. In particular, a concerning kind of failure is that many tests check docker logs and now the output of the services isn't going to the log at all, apparently (e.g. [this check](https://github.com/canonical/rockcraft/blob/main/tests/spread/general/entrypoint-service/task.yaml#L15)) is failing.

I think this is a breaking change to pull on the users - the logs for their rocks will suddenly be empty and they won't know until they check them (presumably because something broke and they need to check the logs)

**Review by cjdcordeiro — CHANGES_REQUESTED** [review]
You'll need to fix the spread tests that expect logs to be coming out by default. I suggest you fix them in a way that 1) in some cases you explicitly pass the `--verbose` to then assert the output, and 2) in other cases there's simply no service output

**Comment by cjdcordeiro** [comment]
> I think this is a breaking change to pull on the users - the logs for their rocks will suddenly be empty and they won't know until they check them (presumably because something broke and they need to check the logs)

Not necessarily a breaking change, but indeed something that is easy to miss. @linostar is giving people a heads-up on Discourse and later on in Matrix

The `--verbose` was introduced in response to an original request from @simskij (Simon please be aware of this incoming change and speak out if it's gonna break something for you).

TBH, this was an oversight because:
 1. for logs, there's the `pebble logs` command that can be used at any time after container deployment
 2. there's the `--verbose` flag that can also be passed at container deployment, but we don't have a `--quiet` flag. So by passing `--verbose` by default, we're not giving a choice to silence the output
 3. in cases where you'd like to do something like `docker run ... exec`, the default verbosity is just polluting the output

**Comment by simskij** [comment]
I might be mistaken, and in that case, please enlighten me, but if im not completely mistaken this change breaks the entire charm workflow, as logs no longer will be picked up in `juju debug-logs` since they get their logs from pebble's `stdout`. I'm also interested in knowing whether this change affects the pebble log forwarding feature at all or whether that will remain unaffected by the change.

**Comment by cjdcordeiro** [comment]
thanks @simskij . that's important.

let me add @tonyandrewmeyer and @benhoyt here, since this is also related to Pebble and the operator framework.

@benhoyt, in short, the idea is that rocks would have pebble run without the --verbose from now on. Seeing @simskij's message above, do you see any way a charmer could enable the `--verbose` at deployment time?

**Comment by benhoyt** [comment]
This is definitely important, however ... when Juju starts the workload containers it starts pebble with these args (code [here](https://github.com/juju/juju/blob/a90328ccf574d4be960b0236a5c21cd030111fd2/caas/kubernetes/provider/application/application.go#L1477-L1484)):

```
/charm/bin/pebble run --create-dirs --hold --http=:<api-port> --verbose
```

So `--verbose` is explicitly specified to K8s in the command args, and as such, I don't think this breaks the Juju logging flow.

In addition, does `juju debug-log` pick up workload container logs now? I don't think it does, just logs from the charm container (and hence Ops framework).

This change also does not affect Pebble's log forwarding feature at all. That's configured in the Pebble config layer and is internal to Pebble.

So I think we're good here as far as charming goes. Unless I'm missing something, it seems to me this would only apply when you're running a Rock via docker at the command line.

That said, I think it's worth testing end-to-end with Juju and a Rock built with this non-verbose `rockcraft` change, just to be sure we're not regressing here!

**Comment by cjdcordeiro** [comment]
Tyvm @benhoyt ! That's useful.

@linostar will take this opportunity to get started with Juju and charms, and give it a try.

But @simskij , it would be paramount to have your appraisal, since you already have proper use cases. On your side, I think the test could be quite simple. Just take an existing rock, and do

```bash
echo '''
FROM <your-rock>
ENTRYPOINT ["pebble", "enter"]
''' | docker build -t <your-quiet-rock> -
```

Then use this new rock in your charm and assess the impact.

**Comment by linostar** [comment]
@benhoyt 
So I built 2 hello-world rocks, one called `hello` that has `--verbose` pebble option, and the other called `hola` without `--verbose` option.
Next, I created a charm for each of those 2 rocks, using the rock as a charm resource. I created a pebble layer in the charms that has a service that just runs the hello binary.
After deploying each charm, the `juju debug-log` output looks similar. The error logged is just because the hello binary don't stay running in the background and exits quickly.

**hello charm debug logs**
```
----- Logs from task 0 -----
2024-04-02T14:23:54Z INFO Most recent service output:
    Hello, world!
2024-04-02T14:23:54Z ERROR cannot start service: exited quickly with code 0
-----
unit-hello-charm-0: 17:23:55 ERROR juju.worker.uniter.operation hook "hello-pebble-ready" (via hook dispatching script: dispatch) failed: exit status 1
unit-hello-charm-0: 17:23:55 ERROR juju.worker.uniter pebble poll failed for container "hello": failed to send pebble-ready event: hook failed
unit-hello-charm-0: 17:23:59 DEBUG unit.hello-charm/0.juju-log ops 2.12.0 up and running.
unit-hello-charm-0: 17:23:59 DEBUG unit.hello-charm/0.juju-log Emitting Juju event hello_pebble_ready.
unit-hello-charm-0: 17:23:59 ERROR unit.hello-charm/0.juju-log Uncaught exception while in charm code:
Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-hello-charm-0/charm/./src/charm.py", line 46, in <module>
    ops.main(HelloCharm)
  File "/var/lib/juju/agents/unit-hello-charm-0/charm/venv/ops/main.py", line 555, in __call__
    return main(charm_class, use_juju_for_storage=use_juju_for_storage)
  File "/var/lib/juju/agents/unit-hello-charm-0/charm/venv/ops/main.py", line 544, in main
    manager.run()
  File "/var/lib/juju/agents/unit-hello-charm-0/charm/venv/ops/main.py", line 520, in run
    self._emit()
  File "/var/lib/juju/agents/unit-hello-charm-0/charm/venv/ops/main.py", line 509, in _emit
    _emit_charm_event(self.charm, self.dispatcher.event_name)
  File "/var/lib/juju/agents/unit-hello-charm-0/charm/venv/ops/main.py", line 143, in _emit_charm_event
    event_to_emit.emit(*args, **kwargs)
  File "/var/lib/juju/agents/unit-hello-charm-0/charm/venv/ops/framework.py", line 352, in emit
    framework._emit(event)
  File "/var/lib/juju/agents/unit-hello-charm-0/charm/venv/ops/framework.py", line 851, in _emit
    self._reemit(event_path)
  File "/var/lib/juju/agents/unit-hello-charm-0/charm/venv/ops/framework.py", line 941, in _reemit
    custom_handler(event)
  File "/var/lib/juju/agents/unit-hello-charm-0/charm/./src/charm.py", line 25, in _on_hello_pebble_ready
    container.replan()
  File "/var/lib/juju/agents/unit-hello-charm-0/charm/venv/ops/model.py", line 2129, in replan
    self._pebble.replan_services()
  File "/var/lib/juju/agents/unit-hello-charm-0/charm/venv/ops/pebble.py", line 1905, in replan_services
    return self._services_action('replan', [], timeout, delay)
  File "/var/lib/juju/agents/unit-hello-charm-0/charm/venv/ops/pebble.py", line 1989, in _services_action
    raise ChangeError(change.err, change)
ops.pebble.ChangeError: cannot perform the following tasks:
- Start service "hello-service" (cannot start service: exited quickly with code 0)
```

**hola charm debug logs**
```
----- Logs from task 0 -----
2024-04-02T14:05:01Z INFO Most recent service output:
    Hello, world!
2024-04-02T14:05:01Z ERROR cannot start service: exited quickly with code 0
-----
unit-hello-charm-0: 17:05:02 ERROR juju.worker.uniter.operation hook "hello-pebble-ready" (via hook dispatching script: dispatch) failed: exit status 1
unit-hello-charm-0: 17:05:02 ERROR juju.worker.uniter pebble poll failed for container "hello": failed to send pebble-ready event: hook failed
unit-hello-charm-0: 17:05:06 DEBUG unit.hello-charm/0.juju-log ops 2.12.0 up and running.
unit-hello-charm-0: 17:05:06 DEBUG unit.hello-charm/0.juju-log Emitting Juju event hello_pebble_ready.
unit-hello-charm-0: 17:05:06 ERROR unit.hello-charm/0.juju-log Uncaught exception while in charm code:
Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-hello-charm-0/charm/./src/charm.py", line 46, in <module>
    ops.main(HelloCharm)
  File "/var/lib/juju/agents/unit-hello-charm-0/charm/venv/ops/main.py", line 555, in __call__
    return main(charm_class, use_juju_for_storage=use_juju_for_storage)
  File "/var/lib/juju/agents/unit-hello-charm-0/charm/venv/ops/main.py", line 544, in main
    manager.run()
  File "/var/lib/juju/agents/unit-hello-charm-0/charm/venv/ops/main.py", line 520, in run
    self._emit()
  File "/var/lib/juju/agents/unit-hello-charm-0/charm/venv/ops/main.py", line 509, in _emit
    _emit_charm_event(self.charm, self.dispatcher.event_name)
  File "/var/lib/juju/agents/unit-hello-charm-0/charm/venv/ops/main.py", line 143, in _emit_charm_event
    event_to_emit.emit(*args, **kwargs)
  File "/var/lib/juju/agents/unit-hello-charm-0/charm/venv/ops/framework.py", line 352, in emit
    framework._emit(event)
  File "/var/lib/juju/agents/unit-hello-charm-0/charm/venv/ops/framework.py", line 851, in _emit
    self._reemit(event_path)
  File "/var/lib/juju/agents/unit-hello-charm-0/charm/venv/ops/framework.py", line 941, in _reemit
    custom_handler(event)
  File "/var/lib/juju/agents/unit-hello-charm-0/charm/./src/charm.py", line 25, in _on_hello_pebble_ready
    container.replan()
  File "/var/lib/juju/agents/unit-hello-charm-0/charm/venv/ops/model.py", line 2129, in replan
    self._pebble.replan_services()
  File "/var/lib/juju/agents/unit-hello-charm-0/charm/venv/ops/pebble.py", line 1905, in replan_services
    return self._services_action('replan', [], timeout, delay)
  File "/var/lib/juju/agents/unit-hello-charm-0/charm/venv/ops/pebble.py", line 1989, in _services_action
    raise ChangeError(change.err, change)
ops.pebble.ChangeError: cannot perform the following tasks:
- Start service "hello-service" (cannot start service: exited quickly with code 0)
```

So basically, `juju debug-log` doesn't seem to log the output of `pebble enter --verbose` anyway.
Let me know if there is another use case you want to test.

**Comment by cjdcordeiro** [comment]
@linostar you probably need to fix your use case, because hello-world isn't really a proper pebble service, thus why Pebble is exiting with an error -> the service doesn't run for long enough. You'll need to make a long-lived noisy service

**Comment by linostar** [comment]
After conferring with each of:
1. Alex Lowe, Charmcraft expert
2. Tom Haddon, juju expert
3. Simon Aronsson and Leon Mintz from Observability team
and based on [Ben Hoyt's comment](https://github.com/canonical/rockcraft/pull/495#issuecomment-1989741567) above, I came to the following conclusions:
- the `--verbose` argument is explicitly passed to pebble in juju
- `juju debug-log` doesn't show any container-related logs, unless `charm.py` explicitly using the logging module to log something coming from the container(s).

I conducted a new experiment with charms that use rock-based containers, as my previous experiment was faulted due to deploying a fast-exiting service in pebble. I deployed 2 charms: one using a base-22.04 rock (so with --verbose), and the other using a base-24.04 rock (so without --verbose). For both charms, they didn't log any container logs as can be seen from the logs below:

charm with `--verbose`:
```
unit-charm-verbose-0: 17:05:14 INFO juju.cmd running containerAgent [3.1.8 810900f47952a1f3835576f57dce2f9d1aef23d0 gc go1.21.9]
unit-charm-verbose-0: 17:05:19 DEBUG unit.charm-verbose/0.juju-log ops 2.12.0 up and running.
unit-charm-verbose-0: 17:05:19 INFO unit.charm-verbose/0.juju-log Running legacy hooks/install.
unit-charm-verbose-0: 17:05:19 DEBUG unit.charm-verbose/0.juju-log ops 2.12.0 up and running.
unit-charm-verbose-0: 17:05:19 DEBUG unit.charm-verbose/0.juju-log Charm called itself via hooks/install.
unit-charm-verbose-0: 17:05:19 DEBUG unit.charm-verbose/0.juju-log Legacy hooks/install exited with status 0.
unit-charm-verbose-0: 17:05:19 DEBUG unit.charm-verbose/0.juju-log Using local storage: not a Kubernetes podspec charm
unit-charm-verbose-0: 17:05:19 DEBUG unit.charm-verbose/0.juju-log Initializing SQLite local storage: /var/lib/juju/agents/unit-charm-verbose-0/charm/.unit-state.db.
unit-charm-verbose-0: 17:05:19 DEBUG unit.charm-verbose/0.juju-log Emitting Juju event install.
unit-charm-verbose-0: 17:05:19 DEBUG unit.charm-verbose/0.juju-log ops 2.12.0 up and running.
unit-charm-verbose-0: 17:05:19 DEBUG unit.charm-verbose/0.juju-log Emitting Juju event leader_elected.
unit-charm-verbose-0: 17:05:20 DEBUG unit.charm-verbose/0.juju-log ops 2.12.0 up and running.
unit-charm-verbose-0: 17:05:20 DEBUG unit.charm-verbose/0.juju-log Emitting Juju event verbose_pebble_ready.
unit-charm-verbose-0: 17:05:20 DEBUG unit.charm-verbose/0.juju-log ops 2.12.0 up and running.
unit-charm-verbose-0: 17:05:20 DEBUG unit.charm-verbose/0.juju-log Emitting Juju event config_changed.
unit-charm-verbose-0: 17:05:21 DEBUG unit.charm-verbose/0.juju-log ops 2.12.0 up and running.
unit-charm-verbose-0: 17:05:21 INFO unit.charm-verbose/0.juju-log Running legacy hooks/start.
unit-charm-verbose-0: 17:05:21 DEBUG unit.charm-verbose/0.juju-log ops 2.12.0 up and running.
unit-charm-verbose-0: 17:05:21 DEBUG unit.charm-verbose/0.juju-log Charm called itself via hooks/start.
unit-charm-verbose-0: 17:05:21 DEBUG unit.charm-verbose/0.juju-log Legacy hooks/start exited with status 0.
unit-charm-verbose-0: 17:05:21 DEBUG unit.charm-verbose/0.juju-log Emitting Juju event start.

```

charm without `--verbose`:
```
unit-charm-quiet-0: 16:17:01 INFO juju.cmd running containerAgent [3.1.8 810900f47952a1f3835576f57dce2f9d1aef23d0 gc go1.21.9]
unit-charm-quiet-0: 16:17:01 INFO juju.cmd.containeragent.unit start "unit"
unit-charm-quiet-0: 16:17:01 INFO juju.worker.upgradesteps upgrade steps for 3.1.8 have already been run.
unit-charm-quiet-0: 16:17:01 INFO juju.worker.probehttpserver starting http server on 127.0.0.1:65301
unit-charm-quiet-0: 16:17:01 INFO juju.api connection established to "wss://controller-service.controller-microk8s.svc.cluster.local:17070/model/b2ed3941-b8c5-4888-8302-ced5984a4cea/api"
unit-charm-quiet-0: 16:17:01 INFO juju.worker.apicaller [b2ed39] "unit-charm-quiet-0" successfully connected to "controller-service.controller-microk8s.svc.cluster.local:17070"
unit-charm-quiet-0: 16:17:01 INFO juju.api cannot resolve "controller-service.controller-microk8s.svc.cluster.local": lookup controller-service.controller-microk8s.svc.cluster.local: operation was canceled
unit-charm-quiet-0: 16:17:01 INFO juju.api connection established to "wss://10.152.183.225:17070/model/b2ed3941-b8c5-4888-8302-ced5984a4cea/api"
unit-charm-quiet-0: 16:17:01 INFO juju.worker.apicaller [b2ed39] "unit-charm-quiet-0" successfully connected to "10.152.183.225:17070"
unit-charm-quiet-0: 16:17:01 INFO juju.worker.migrationminion migration phase is now: NONE
unit-charm-quiet-0: 16:17:01 INFO juju.worker.logger logger worker started
unit-charm-quiet-0: 16:17:01 WARNING juju.worker.proxyupdater unable to set snap core settings [proxy.http= proxy.https= proxy.store=]: exec: "snap": executable file not found in $PATH, output: ""
unit-charm-quiet-0: 16:17:17 DEBUG unit.charm-quiet/0.juju-log ops 2.12.0 up and running.
unit-charm-quiet-0: 16:17:17 INFO unit.charm-quiet/0.juju-log Running legacy hooks/install.
unit-charm-quiet-0: 16:17:18 DEBUG unit.charm-quiet/0.juju-log ops 2.12.0 up and running.
unit-charm-quiet-0: 16:17:18 DEBUG unit.charm-quiet/0.juju-log Charm called itself via hooks/install.
unit-charm-quiet-0: 16:17:18 DEBUG unit.charm-quiet/0.juju-log Legacy hooks/install exited with status 0.
unit-charm-quiet-0: 16:17:18 DEBUG unit.charm-quiet/0.juju-log Using local storage: not a Kubernetes podspec charm
unit-charm-quiet-0: 16:17:18 DEBUG unit.charm-quiet/0.juju-log Initializing SQLite local storage: /var/lib/juju/agents/unit-charm-quiet-0/charm/.unit-state.db.
unit-charm-quiet-0: 16:17:18 DEBUG unit.charm-quiet/0.juju-log Emitting Juju event install.
unit-charm-quiet-0: 16:17:18 DEBUG unit.charm-quiet/0.juju-log ops 2.12.0 up and running.
unit-charm-quiet-0: 16:17:18 DEBUG unit.charm-quiet/0.juju-log Emitting Juju event leader_elected.
unit-charm-quiet-0: 16:17:18 DEBUG unit.charm-quiet/0.juju-log ops 2.12.0 up and running.
unit-charm-quiet-0: 16:17:18 DEBUG unit.charm-quiet/0.juju-log Emitting Juju event quiet_pebble_ready.
unit-charm-quiet-0: 16:17:19 DEBUG unit.charm-quiet/0.juju-log ops 2.12.0 up and running.
unit-charm-quiet-0: 16:17:19 DEBUG unit.charm-quiet/0.juju-log Emitting Juju event config_changed.
unit-charm-quiet-0: 16:17:19 DEBUG unit.charm-quiet/0.juju-log ops 2.12.0 up and running.
unit-charm-quiet-0: 16:17:19 INFO unit.charm-quiet/0.juju-log Running legacy hooks/start.
unit-charm-quiet-0: 16:17:19 DEBUG unit.charm-quiet/0.juju-log ops 2.12.0 up and running.
unit-charm-quiet-0: 16:17:19 DEBUG unit.charm-quiet/0.juju-log Charm called itself via hooks/start.
unit-charm-quiet-0: 16:17:19 DEBUG unit.charm-quiet/0.juju-log Legacy hooks/start exited with status 0.
unit-charm-quiet-0: 16:17:19 DEBUG unit.charm-quiet/0.juju-log Emitting Juju event start.
```

Furthermore, we already deployed this `--verbose` drop change in Rockcraft 1.3.x, but it applies only to rocks with base 24.04 or above. So far, no issues were reported.

Finally, [juju documentation](https://juju.is/docs/sdk/interact-with-pebble#heading--view-service-logs) confirms that `--verbose` is passed to juju's pebble service.

We aim to merge the current PR in Madrid, so people can have time to experiment this change with 24.04 base.

**Comment by benhoyt** [comment]
@linostar Any status update on this? Do you folks think you'll be merging this PR soon?

**Comment by linostar** [comment]
> @linostar Any status update on this? Do you folks think you'll be merging this PR soon?

We're waiting on @simskij to run some tests with a rockcraft snap that has `--verbose` dropped from rock's entrypoint.
Once Simon confirms all is well we'll merge this PR.

**Comment by linostar** [comment]
@cjdcordeiro @tigarmo 

Thanks to Jose Masson from the Observability team, following the instructions of Simon, the test was done using the snap build from PR #572 (which has the --verbose argument dropped for all bases).

Here are the test and the results:


-----------------------
**Jose:**

Well, using this rockcraft version:

```
snap install --classic --channel=edge/pr-572 rockcraft
```

I packed this rock: https://github.com/canonical/prometheus-rock/tree/main/2.50.1
Then I used that rock to deploy this prometheus charm:  https://github.com/canonical/prometheus-k8s-operator/

```
juju deploy ./*.charm prom --resource prometheus-image=docker.io/josemasson/prometheus:2.50.1 --trust
```

Once the charm was deployed I verified that workloads logs are there:

```
╭─ubuntu@charm-dev-34 ~/repos/prometheus-k8s-operator ‹main› [microk8s:prom]
╰─$ juju ssh --container prometheus prom/0 pebble logs                                                                                   1 ↵
2024-05-30T20:56:54.890Z [prometheus] ts=2024-05-30T20:56:54.890Z caller=main.go:608 level=info msg="Starting Prometheus Server" mode=server version="(version=2.50.1, branch=HEAD, revision=8c9b0285360a0b6288d76214a75ce3025bce4050)"
2024-05-30T20:56:54.890Z [prometheus] ts=2024-05-30T20:56:54.890Z caller=main.go:613 level=info build_context="(go=go1.20.14, platform=linux/amd64, user=root@rockcraft-prometheus-on-amd64-for-amd64-31498686, date=20240517-08:50:09, tags=netgo,builtinassets,stringlabels)"
2024-05-30T20:56:54.890Z [prometheus] ts=2024-05-30T20:56:54.890Z caller=main.go:614 level=info host_details="(Linux 5.15.0-102-generic #112-Ubuntu SMP Tue Mar 5 16:50:32 UTC 2024 x86_64 prom-0 (none))"
2024-05-30T20:56:54.890Z [prometheus] ts=2024-05-30T20:56:54.890Z caller=main.go:615 level=info fd_limits="(soft=65536, hard=65536)"
2024-05-30T20:56:54.890Z [prometheus] ts=2024-05-30T20:56:54.890Z caller=main.go:616 level=info vm_limits="(soft=unlimited, hard=unlimited)"
2024-05-30T20:56:54.891Z [prometheus] ts=2024-05-30T20:56:54.891Z caller=web.go:565 level=info component=web msg="Start listening for connections" address=0.0.0.0:9090
2024-05-30T20:56:54.891Z [prometheus] ts=2024-05-30T20:56:54.891Z caller=main.go:1118 level=info msg="Starting TSDB ..."
2024-05-30T20:56:54.893Z [prometheus] ts=2024-05-30T20:56:54.893Z caller=tls_config.go:313 level=info component=web msg="Listening on" address=[::]:9090
2024-05-30T20:56:54.893Z [prometheus] ts=2024-05-30T20:56:54.893Z caller=tls_config.go:316 level=info component=web msg="TLS is disabled." http2=false address=[::]:9090
2024-05-30T20:56:54.893Z [prometheus] ts=2024-05-30T20:56:54.893Z caller=head.go:610 level=info component=tsdb msg="Replaying on-disk memory mappable chunks if any"
2024-05-30T20:56:54.894Z [prometheus] ts=2024-05-30T20:56:54.894Z caller=head.go:692 level=info component=tsdb msg="On-disk memory mappable chunks replay completed" duration=3.467µs
2024-05-30T20:56:54.894Z [prometheus] ts=2024-05-30T20:56:54.894Z caller=head.go:700 level=info component=tsdb msg="Replaying WAL, this may take a while"
2024-05-30T20:56:54.894Z [prometheus] ts=2024-05-30T20:56:54.894Z caller=head.go:771 level=info component=tsdb msg="WAL segment loaded" segment=0 maxSegment=0
```
-----------------------

The rockcraft service result (prometheus) are still seen in the workload logs, as it can be seen above.
I believe that with this we can be bring this PR to a happy ending.

If no further controversy or extra testing are required, I will proceed to update the branch so it becomes ready for merging.

**Comment by cjdcordeiro** [comment]
thanks for the update! 

Yes please, let's go for it.

As an extra cautionary step, we can then leave the snap in `edge` for a couple of weeks before promoting, to make sure no one else is affected

**Comment by linostar** [comment]
@tigarmo @cjdcordeiro Ready for review.

File: tests/spread/foreign/big/task.yaml [hunk 1]
```diff
@@ -27,7 +27,9 @@ execute: |
   docker images
 
   # Check the rock's output
-  docker run --rm big:latest | MATCH "/tmp"
+  id=$(docker run -d big:latest --verbose)
```
Comment (cjdcordeiro) [inline tests/spread/foreign/big/task.yaml:30]: Why not keep the `--rm` and do the assertion in a single line, like before?

File: tests/spread/rockcraft/entrypoint-service/task.yaml [hunk 2]
```diff
@@ -10,11 +10,11 @@ execute: |
...
-  test "$(docker inspect "$id" -f '{{json .Config.Entrypoint}}')" = '["/bin/pebble","enter","--verbose","--args","test-service"]'
+  test "$(docker inspect "$id" -f '{{json .Config.Entrypoint}}')" = '["/bin/pebble","enter","--args","test-service"]'
   test "$(docker inspect "$id" -f '{{json .Config.Cmd}}')" = '["foo"]'
-  grep_docker_log "$id" "foo"
+  docker logs "$id" 2>&1 | grep "foo"
   docker rm -f "$id"
   id=$(docker run -d entrypoint-service-test bar)
   test "$(docker inspect "$id" -f '{{json .Config.Cmd}}')" = '["bar"]'
-  grep_docker_log "$id" "bar"
+  docker logs "$id" 2>&1 | grep "bar"
```
Comment (cjdcordeiro) [inline tests/spread/rockcraft/entrypoint-service/task.yaml:19]: I think it would be nice to have one of these tests run with `\; --verbose`

**Review by tigarmo — APPROVED** [review]

**Review by cjdcordeiro — APPROVED** [review]
lgtm thanks

**Review by cjdcordeiro — CHANGES_REQUESTED** [review]
I see a test is still failing. pls re-request review once it passes

**Comment by linostar** [comment]
> I see a test is still failing. pls re-request review once it passes

The error in question is unrelated and is due to fedora 37 not available for workers any more. I'm trying to fix that in PR https://github.com/canonical/rockcraft/pull/580.

**Comment by linostar** [comment]
All related spread tests seem to pass, but some tests are being killed to timeout due to some slowness in spread workers (the exception is the fedora worker addressed in PR #580). Even when after re-running the spread tests several times, different tests fail every time.

**Review by cjdcordeiro — APPROVED** [review]

**Comment by linostar** [comment]
@tigarmo can you please merge this before something else merged in main and cause a new conflict?

**Comment by claudiubelu** [comment]
Hello.

I was wondering why some rocks randomly had service logs and other did not. It seems that the difference was the rockcraft version being used, and I eventually found the reason: this PR.

This is a bit late at this point, but I'd like the ``--verbose`` flag to be back, or some ``rockcraft.yaml`` option to add it (``False`` by default).

We are using ``rockcraft`` to build rocks meant to be used in Kubernetes directly, used in various Helm charts for various operators, all without any juju involved at any stage. Because the entrypoint is always Pebble, when you execute ``kubectl logs``, you'd always get the Pebble logs instead of the actual service you'd actually want to see. Even while creating and testing our rocks, we've had plenty of mishaps because of this. For example, in one case we had a rock built and was testing in Kubernetes, the service was starting, the liveness probe was working, however, the Pod would eventually die and end up in a ``CrashLoopBackOff`` state... and we'd have no way of actually knowing why that happened in the first place. In Kubernetes, you can run ``kubectl logs`` to see the Pod logs, even if that Pod is not ``Running``. However, in our case we would only see the Pebble logs, which were of no use. Running ``kubectl exec`` into the Pod to run ``pebble logs`` was not option because the Pod has to be ``Running`` to do that. Changing the Pebble service's ``on-failure`` / ``on-success`` options would be a very time-consuming option to debug this further (the rock would have to be rebuilt and pushed to a registry), and it wouldn't always work anyways (some Pods may have Readiness probes, which, if if they fail, the Pod is killed). Instead, we edited the daemonset's / deployment's command to bypass Pebble entirely, just so we can debug this issue further. But in terms of user experience, something like this may not be the first thing that would come to mind for different users that would try to use that image, they might not even know what a rock or Pebble is, all they would see would be some confusing logs from some unknown service that seems to be running instead of what they were expecting to see running, especially since the rocks we're building are meant to be drop-in replacements for other images (or at least as close as it can be). This wouldn't have been the case if the rocks were ``--verbose`` in the first place.

I see a lot of comments here are related to ``rockcraft`` being used in juju-related cases, that juju is already adding this flag anyways. IMO, that shouldn't have been a reason to remove this flag, especially since someone requested it in the first place, but an option to enable / disable it would have been great.

I am willing to work on adding that option to ``rockcraft``, especially since I've been looking into ``rockcraft`` internals for a bit.

Best regards,

Claudiu

**Comment by cjdcordeiro** [comment]
Hi Claudiu,

thank you for the detailed feedback and for sharing your use case.

This was indeed an intentional change, that despite its relation to Juju, was meant to fix what we believed to be an undesired default behavior -> being too verbose.

While it might sound obvious that one always wants their service to print out its logs to STDOUT by default, in the case of a rock we must consider other factors:
 - we're using an init system - Pebble - and thus, as you correctly mentioned, some logs would be coming from Pebble while others come from services
 - having `--verbose` by default might sound obvious for a single service, but can quickly become too noisy if one has multiple services
 - using `--verbose` with Pebble commands other than `run` would also cause unnecessary noise.

The goal here is that one can always add `--verbose` at runtime, but you cannot "silence" Pebble if `--verbose` is hardcoded into the entrypoint.

In the near future, we might be considering some new features that may be useful for specific logging cases, like [Log Forwarding](https://canonical-pebble.readthedocs-hosted.com/en/latest/reference/log-forwarding/) and the ability to add more control, at build time, to what the OCI cmd should be.

In the meantime, adding `-v` at runtime is the only way to reproduce the old behaviour.
